### PR TITLE
Generating ASR for templates

### PR DESCRIPTION
**Pull request by Oshanath**
* A symbol table was added to the Template ASR.
* The interface in the example had to be removed. Will add it later.

### CODE REVIEW

File: examples/swap_m.f90 [hunk 1]
```diff
@@ -10,12 +10,9 @@ module swap_m
         type :: T
         end type
 
-        interface swap
-            module procedure swap_
-        end interface
```
Comment (certik) [inline examples/swap_m.f90]: Why was this interface removed?

Comment (Oshanath) [reply examples/swap_m.f90]: I removed it to first get it working and then add the functionality later.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 2]
```diff
@@ -831,6 +831,11 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         for (size_t i=0; i<x.n_contains; i++) {
             visit_program_unit(*x.m_contains[i]);
         }
+        for (size_t i=0; i<x.n_decl; i++) {
+            visit_unit_decl2(*x.m_decl[i]);
+        }
+        // We have to visit unit_decl_2 because in the example, the Template is directly inside the module and 
+        // Template is a unit_decl_2
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:855]: Yes, good point. The body visitor must  visit all function bodies, and there are some functions "hidden" inside the template.

**Comment by certik** [comment]
Very good, this looks excellent.

**Comment by certik** [comment]
Now you need to update the AST test for "swap", since you modified the source code (see my comment above). Then let's add an ASR test for this.

After that, I think we are done, and we'll work on instantiation.

**Comment by Oshanath** [comment]
I think there are a few more things to be done.
- I have to identify `TypeParameter`s.
- I have to populate the `TypeParameter`s of the `Template`.
- I have to return a `TemplatedFunction` if the function uses `TypeParameter`s.

**Comment by certik** [comment]
Yes. But if you create the tests, then everything will pass and be "done", then you iteratively fix all the things you listed by implementing them and updating test for each. Which is how everything else in LFortran/LPython is also developed.

**Comment by Oshanath** [comment]
I see. I will do it in that manner.

**Comment by Oshanath** [comment]
@certik I updated the AST test and added an ASR test.

**Comment by czgdp1807** [comment]
Well, if merging this into main branch will cause any issue with compilation of already existing Fortran code (integration_tests, tests) with LFortran then we shouldn't merge this. Reason it would cause problems with addition of other features. However, if merging this will keep the workflow of other developers in our team unaffected then feel free to go ahead and merge. Updating syntax for templates is not a concern since we are not offering anything stable to our users as of now. LFortran is already changing daily so syntax updates should be expected and should be normal.

**Comment by certik** [comment]
As far as I know, there is no breakage to any existing Fortran code. If we find one in the future, we'll have to disable or change the template syntax of course. The only possible downside is that we might want to support some different syntax in the future, either for templates or for something else (possibly standardized in the future!) that would conflict with this. In that case, since currently "templates" are currently alpha, we would simply change the syntax for templates. Later, if we decide to promote templates to beta, we might need to evaluate more carefully. It seems an "alpha" support for a syntax is not a big deal. So we'll get it to a stage when we can merge.

**Comment by certik** [comment]
This was merged as part of https://github.com/lfortran/lfortran/pull/90.

**Pull request closed without merging**